game-core/src/main/java/games/strategy/triplea/attachments/FireTriggerParams.java

```diff
@@ -0,0 +1,75 @@
+package games.strategy.triplea.attachments;
+
+
+/**
+ * Parameter object for fire trigger methods, first and foremost in {@link TriggerAttachment}.
+ */
+public class FireTriggerParams {
+
+  /**
+   * Filter value for before or after a given step.
+   *
+   * <p>
+   * See {@link #testWhen}, documentation for trigger attachment option "when" as well as
+   * {@link AbstractTriggerAttachment#whenOrDefaultMatch} .
+   * </p>
+   */
+  public final String beforeOrAfter;
+
+  /**
+   * Filter value for a specific step/phase (like "russianPolitics" step).
+   *
+   * <p>
+   * See {@link #testWhen}, documentation for trigger attachment option "when" as well as
+   * {@link AbstractTriggerAttachment#whenOrDefaultMatch} .
+   * </p>
+   */
+  public final String stepName;
+
+  /**
+   * TODO: Document.
+   *
+   * <p>
+   * See documentation for trigger attachment option "uses".
+   * </p>
+   */
+  public final boolean useUses;
+
+  /**
+   * TODO: Document.
+   *
+   * <p>
+   * See documentation for trigger attachment option "uses".
+   * </p>
+   */
+  public final boolean testUses;
+
+  /**
+   * TODO: Document.
+   *
+   * <p>
+   * See documentation for trigger attachment option "chance".
+   * </p>
+   */
+  public final boolean testChance;
+
+  /**
+   * Whether to filter for "when".
+   *
+   * <p>
+   * See documentation for trigger attachment option "when" as well as
+   * {@link AbstractTriggerAttachment#whenOrDefaultMatch} .
+   * </p>
+   */
+  public final boolean testWhen;
+
+  public FireTriggerParams(final String beforeOrAfter, final String stepName,
+      final boolean useUses, final boolean testUses, final boolean testChance, final boolean testWhen) {
+    this.beforeOrAfter = beforeOrAfter;
+    this.stepName = stepName;
+    this.useUses = useUses;
+    this.testUses = testUses;
+    this.testChance = testChance;
+    this.testWhen = testWhen;
+  }
```

Comment:
I'm not entirely sure about this class. Using public non-static fields is not really common for java.
You _could_ use lombok to generate getters for you but I'm not entirely sure either if this actually is better.

Reply:
I agree. I did actually try to use Lombok, but the "getter" for the boolean fields ended up with `is` prepended, and that was verbose and did not look quite right (though I could probably make the variable names I used for `FireTriggerParams` in the various places more concise). I looked into renaming the "accessors", but it seems to be an experimental feature in Lombok: https://projectlombok.org/features/experimental/Accessors .

Thinking about it, I might be able to use Lombok, and then disable automatic generation of "getters" and create them myself, such that `equals` and `hashCode` is generated while still getting decently named "getters".

About a kind of data class like this, you are fully right that it is non-idiomatic for Java in multiple regards, though I believe it makes somewhat sense in this case, maybe if I improved it further. The trouble with not using "getters" in Java is that you then cannot modify how it is accessed later. In languages such as (I believe) C# and Scala (and Kotlin?), you can add "getters" later, maybe including reg. binary compatibility, which makes that aspect much less of an issue.

Likewise for the immutable, final access, that is non-typical for Java, for multiple reasons. In many cases, encapsulation and protecting access to members make a lot of sense, since in many ways the specific internal implementation is not meant to be a part of the interface or be used. In other cases, it may make sense to make the fields directly accessible (though maybe through a "getter"), generally if the fields are meant to be a part of the interface. This is still somewhat verbose in Java unless you use Lombok or similar. There are investigations in regards to maybe adding something in this direction to the Java language in the future by Brian Goetz and others: http://cr.openjdk.java.net/~briangoetz/amber/datum.html .

Reply:
I have just looked into creating "getters" named for instance `testChance`, since `isTestChance` seemed off. But that looks off in the following line, since there is already a method on `AbstractTriggerAttachment` also named `testChance`:

`if (fireTriggerParams.testChance() && !t.testChance(bridge)) {`.

I am not certain about the approach. Maybe keep the fields `public` and `final`, disable "getters", and use `Value` to auto-generate `equals`, `hashCode` and `toString`?